File: src/InMemoryRepository.java
import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<ID, E>{

    private Map<ID, E> entities = new HashMap<>();
    @Override
    public E findOne(ID id) {
        if(id == null){
            throw new IllegalArgumentException("Null id");
        }
        return entities.get(id);
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E save(E entity) {
        if(entity == null){
            throw new IllegalArgumentException("Null entity");
        }
        E existing_entity = entities.get(entity.getId());
        if (existing_entity != null)
            return existing_entity;
        entities.put(entity.getId(), entity);
        return null;
    }

    @Override
    public E delete(ID id) {
        if(id == null){
            throw new IllegalArgumentException("Null id");
        }
        return entities.remove(id);
    }

    @Override
    public E update(E entity) {
        if(entity == null){
            throw new IllegalArgumentException("Null entity");
        }
        E existing_entity = entities.get(entity.getId());
        if (existing_entity == null)
            return entity;
        entities.put(entity.getId(), entity);
        return null;
    }
}

File: SEMESTRUL 3/METODE AVANSATE DE PROGRAMARE/SEMINAR/S11/ConsoleApp1/ConsoleApp1/InMemoryRepository.cs
namespace ConsoleApp1;

public class InMemoryRepository<ID, E> : IRepository<ID, E> where E : Entity<ID>
{
    protected IValidator<E> Validator;
    protected IDictionary<ID, E> Entities = new Dictionary<ID, E>();

    public InMemoryRepository(IValidator<E> validator)
    {
        Validator = validator;
    }

    public IEnumerable<E> FindAll()
    {
        return Entities.Values.ToList();
    }

    public E Save(E entity)
    {
        if (entity == null)
        {
            throw new ArgumentNullException("Entitatea nu poate fi null.");
        }

        Validator.Validate(entity);

        if (Entities.ContainsKey(entity.id))
        {
            return entity;
        }

        Entities[entity.id] = entity;
        return default(E);
    }

    public E Delete(ID id)
    {
        if (Entities.ContainsKey(id))
        {
            var entity = Entities[id];
            Entities.Remove(id);
            return entity;
        }

        return default(E);
    }

    public E Update(E entity)
    {
        if (entity == null)
        {
            throw new ArgumentNullException("Entitatea nu poate fi null.");
        }

        Validator.Validate(entity);

        if (!Entities.ContainsKey(entity.id))
        {
            return default(E);
        }
        Entities[entity.id] = entity;
        return default(E);
    }

    public E FindOne(ID id)
    {
        Entities.TryGetValue(id, out var entity);
        return entity;
    }
}